### DISCUSSION TITLE
How to Split Train and Test Dataset in DeepXDE?

### DISCUSSION THREAD

**Discussion by nanas5150** [lululxvi/deepxde#1801] [Q&A]
I'm new to DeepXDE and Physics-Informed Neural Networks (PINNs), and I'm trying to figure out how to properly split my dataset into training and testing sets within DeepXDE.

Could someone please guide me through the process? A simple example or any references to the relevant documentation would be extremely helpful.

Thanks in advance for your help!

**Comment by vl-dud** [comment]
If we are talking about tabular data, then you can use `train_test_split` from scikit-learn

**Reply by nanas5150** [reply]
Thanks for your reply. If you don' mind. can you please guide a simple example of how to integrate train_test_split with DeepXDE? Specifically, I would appreciate an example that shows how to use the split data within a DeepXDE workflow as in inverse problem diffusion reaction below.

def gen_traindata():
    data = np.load("../dataset/reaction.npz")
    t, x, ca, cb = data["t"], data["x"], data["Ca"], data["Cb"]
    X, T = np.meshgrid(x, t)
    X = np.reshape(X, (-1, 1))
    T = np.reshape(T, (-1, 1))
    Ca = np.reshape(ca, (-1, 1))
    Cb = np.reshape(cb, (-1, 1))
    return np.hstack((X, T)), Ca, Cb

kf = dde.Variable(0.05)
D = dde.Variable(1.0)

def pde(x, y):
    ca, cb = y[:, 0:1], y[:, 1:2]
    dca_t = dde.grad.jacobian(y, x, i=0, j=1)
    dca_xx = dde.grad.hessian(y, x, component=0, i=0, j=0)
    dcb_t = dde.grad.jacobian(y, x, i=1, j=1)
    dcb_xx = dde.grad.hessian(y, x, component=1, i=0, j=0)
    eq_a = dca_t - 1e-3 * D * dca_xx + kf * ca * cb ** 2
    eq_b = dcb_t - 1e-3 * D * dcb_xx + 2 * kf * ca * cb ** 2
    return [eq_a, eq_b]

def fun_bc(x):
    return 1 - x[:, 0:1]

def fun_init(x):
    return np.exp(-20 * x[:, 0:1])

geom = dde.geometry.Interval(0, 1)
timedomain = dde.geometry.TimeDomain(0, 10)
geomtime = dde.geometry.GeometryXTime(geom, timedomain)

bc_a = dde.icbc.DirichletBC(
    geomtime, fun_bc, lambda _, on_boundary: on_boundary, component=0
)
bc_b = dde.icbc.DirichletBC(
    geomtime, fun_bc, lambda _, on_boundary: on_boundary, component=1
)
ic1 = dde.icbc.IC(geomtime, fun_init, lambda _, on_initial: on_initial, component=0)
ic2 = dde.icbc.IC(geomtime, fun_init, lambda _, on_initial: on_initial, component=1)

observe_x, Ca, Cb = gen_traindata()
observe_y1 = dde.icbc.PointSetBC(observe_x, Ca, component=0)
observe_y2 = dde.icbc.PointSetBC(observe_x, Cb, component=1)

data = dde.data.TimePDE(
    geomtime,
    pde,
    [bc_a, bc_b, ic1, ic2, observe_y1, observe_y2],
    num_domain=2000,
    num_boundary=100,
    num_initial=100,
    anchors=observe_x,
    num_test=50000,
)
net = dde.nn.FNN([2] + [20] * 3 + [2], "tanh", "Glorot uniform")

model = dde.Model(data, net)
model.compile("adam", lr=0.001, external_trainable_variables=[kf, D])
variable = dde.callbacks.VariableValue([kf, D], period=1000, filename="variables.dat")
losshistory, train_state = model.train(iterations=80000, callbacks=[variable])

**Reply by vl-dud** [reply]
Is a test dataset really needed here? Test cases are generated automatically, see parameter `num_test`

**Reply by nanas5150** [reply]
I have an experiment data for my case study for inverse problem. Is it necessary to split the dataset for train or test? Or in Deepxde it is executed differently?
Also what exactly num_domain=2650, num_boundary=500, num_test=1350 do in the code? Appreciate your kind advise.

def PINN(x,y):
  Ca, Cb,Cc, Cd = y[:, 0:1], y[:, 1:2], y[:, 2:3], y[:, 3:4]
  r = k * Ca * Cb
  dCa_dt = dde.grad.jacobian(y, x, i=0)
  dCb_dt = dde.grad.jacobian(y, x, i=1)
  dCc_dt = dde.grad.jacobian(y, x, i=2)
  dCd_dt = dde.grad.jacobian(y, x, i=3)

  return [dCa_dt + r, dCb_dt + r, dCc_dt - r, dCd_dt - r]

def boundary(_, on_initial):
    return on_initial

#unknown parameter
k = dde.Variable(0.0)

#experiment dataset
observe_Ca = dde.icbc.PointSetBC(tau.reshape(-1, 1), Ca.reshape(-1, 1), component = 0)
observe_Cc = dde.icbc.PointSetBC(tau.reshape(-1, 1), Cc.reshape(-1, 1), component = 2)

geom = dde.geometry.TimeDomain(0, 50)

ic1 = dde.icbc.IC(geom, lambda x: 0.2122, boundary, component=0)
ic2 = dde.icbc.IC(geom, lambda x: 0.4244, boundary, component=1)

data = dde.data.PDE(geom, PINN, [observe_Ca, observe_Cc, ic1, ic2], num_domain=2650, num_boundary=500, num_test=1350)
